### ISSUE TITLE
Improvements

### ISSUE THREAD

**Issue by kiner-shah** [HotelSierraWhiskey/bonsai#6]
I had a brief look and I feel there can be many improvements that can be done. If you want I can help a bit.
- code refactoring
- supporting CMake,
- adding tests and code coverage, etc.

**Comment by HotelSierraWhiskey** [comment]
Testing your code properly will not be possible without hardware, but feel free to submit a PR and I'll take a look.

**Comment by kiner-shah** [comment]
- Bonsai class has API functions and other functions. Are those functions supposed to be public or are they for internal use? Also, they seem to be unimplemented for now.
- Bonsai class has few private members whose meaning is unclear - nvm, part, etc.
- There is a need to add documentation for description of classes, structs, functions and their behavior. Can be added as Doxygen style comments.
- Maybe can put all classes, structs, functions inside a namespace, say `bonsai_fs`.
- The repo structure can be improved, public header files in a include folder and source files, private header files inside src folder.
- Need to add CMakeLists.txt for CMake support.

In order for me to work on this, first 3 points need some clarity. Maybe add some detailed documentation for those.

**Comment by HotelSierraWhiskey** [comment]
Regarding your first question, I had just finished ironing out the bottom layer primative functions before stepping away from this project for a while. The only top level API calls that are currently implemented are `create_file` and `delete_file`. `move_file`, `read_file`, and `write_to_file` need to be implemented (also, that's not the correct signature for `read_file`, it should write to a provided buffer). I remember thinking that the main primitive calls (`get`, `put`, `del`, `mov`) should be public as well. 

As for your second question. The `nvm` member is the non-volatile memory controller driver that comes with Bonsai. It's responsible for doing the actual flash memory operations (namely, page writes and erases) on the chip. The `part` member is retrieved from the `part_table` based on what we read in the Device Service Unit registers (where device ID information is stored). Besides being useful for debugging, the part keeps us honest about how much flash memory is available on the chip, and where the absolute upper bounds of that memory are. 

It was my plan to expose hardware on a linux server that contributers could SSH into to test their code on actual hardware, but life got in the way. I do plan on doing this in the future. This would make testing your firmware actually possible. I have new hardware designed and ready to order. I know there's a need for more documentation, and I'm a big fan of doxygen, so if you want to set that up, feel free. I don't know how familiar you are with ARM Cortex M0+/ SAMC/ SAMD you are, but if you feel like taking a crack at documentation, you're more than welcome. Let me know if you have any more questions, and thanks for taking an interest. This was my first attempt at writing a file system, and I would do it totally differently if I were to start again, but that's life I guess.

**Comment by kiner-shah** [comment]
Yeah I am not familiar with ARM Cortex processors, but I can help with other aspects if required. 
I will try when I get time to add basic Doxygen support and submit a PR. Maybe I will also include CMake support then.

**Comment by kiner-shah** [comment]
I did some minor changes and opened a PR. Please check.
